csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs

```diff
@@ -0,0 +1,195 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using FlatBuffers;
+using System;
+using System.Buffers.Binary;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Apache.Arrow.Ipc
+{
+    internal class ArrowFileReaderImplementation : ArrowStreamReaderImplementation
+    {
+        public bool IsFileValid { get; private set; }
+
+        /// <summary>
+        /// When using GetNextRecordBatch this value 
+        /// is to remember what index is next
+        /// </summary>
+        private int _recordBatchIndex;
+
+        /// <summary>
+        /// Notes what byte position where the footer data is in the stream
+        /// </summary>
+        private int _footerStartPostion;
+
+        private ArrowFooter _footer;
+
+        public ArrowFileReaderImplementation(Stream stream) : base(stream)
+        {
+        }
+
+        public async Task<int> RecordBatchCountAsync()
+        {
+            if (!HasReadSchema)
+            {
+                await ReadSchemaAsync();
+            }
+
+            return _footer.RecordBatchCount;
+        }
+
+        protected override async Task ReadSchemaAsync()
+        {
+            if (HasReadSchema)
+            {
+                return;
+            }
+
+            await ValidateFileAsync();
+
+            var bytesRead = 0;
+            var footerLength = 0;
+
+            await Buffers.RentReturnAsync(4, async (buffer) =>
+            {
+                BaseStream.Position = BaseStream.Length - ArrowFileConstants.Magic.Length - 4;
+
+                bytesRead = await BaseStream.ReadAsync(buffer, 0, 4);
```

Comment:
Ideally this would wrap the `byte[]` into a `Memory<byte>` in order to use the newer `ReadAsync` overload on `Stream`.  The reason for that is that newer overload returns a `ValueTask<int>` rather than ` Task<int>`, which has two benefits:
1) It allows for the stream to avoid allocating a `Task<int>` object just to return the `Int32` value when the operation completes synchronously.  Most stream implementations will benefit from this, including `FileStream`.
2) It allows for the stream implementation to pool an object that backs the `ValueTask<int>` and thus have amortized-zero-allocation even for asynchronously completing operations.  `NetworkStream`, for example, does this.

Reply:
I agree, but I am not planning on making major changes to this code path in this PR. 

I plan on making this code path better with [ARROW-4503: ArrowStreamReader allocates and copies data excessively](https://issues.apache.org/jira/browse/ARROW-4503)